core/katago/katago.go

```diff
@@ -130,102 +140,84 @@ func (an *Analyzer) readCollectedResult(analysisID string) *AnalysisList {
-						glog.Infof("resultCollector: Found sufficent results to return for %s", res.ID)
-						resultSlice := resultStorage[res.ID]
-						an.writeCollectedResult(res.ID, &resultSlice)
-
-						delete(resultStorage, res.ID)
-
-					}
-				} else {
-					glog.Warningf("resultCollector: Expected to find number of results for analysis run, but did not. Run ID: %s", res.ID)
-				}
-			}
+			an.processKatagoOutput(output, res)
 		}
 	}
 }
 
-// Start the Katago Analyzer and make it ready to receive input
-func (an *Analyzer) Start() error {
-	glog.Info("Starting katago analyzer")
-	analyzerSingleton.bootWait.Add(1)
+// processKatagoOutput takes the raw string-output from katago and decides what
+// to do with it.
+func (an *Analyzer) processKatagoOutput(output string, resultStorage map[string]AnalysisList) {
+	if strings.Contains(output, "error") {
+		glog.Warningf("resultCollector: Katago had an error: %v", output)
+		return
+	}
 
-	go an.resultCollector()
-	err := an.startAnalyzerIO()
+	glog.V(3).Infof("Collected %s", output)
+	res, err := ParseAnalysis(output)
 	if err != nil {
-		return err
+		glog.Warningf("resultCollector: Error decoding analysis: %s", output)
 	}
-	err = an.cmd.Start()
-	if err != nil {
-		return err
+
+	// Add the current result to the AnalysisList
+	al := resultStorage[res.ID]
+	resultStorage[res.ID] = append(al, res)
+
+	// If we've collected all the results for this analysis run, notify the caller.
+	sizeLookup, ok := an.analysisSizeMap.Load(res.ID)
+
+	if !ok {
+		glog.Warningf("resultCollector: Expected to find number of results for analysis run, but did not. Run ID: %s", res.ID)
+		return
 	}
-	analyzerSingleton.bootWait.Wait()
 
-	glog.Info("Katago Startup Complete")
-	return nil
+	expectedNumResults := sizeLookup.(int)
+	actualNumResults := len(resultStorage[res.ID])
+	glog.V(3).Infof("resultCollector: found %d/%d expected analysis results for %s", actualNumResults, expectedNumResults, res.ID)
+	if actualNumResults >= expectedNumResults {
+		glog.V(3).Infof("resultCollector: found sufficent results to return for %s", res.ID)
+		resultSlice := resultStorage[res.ID]
+		an.writeCollectedResult(res.ID, &resultSlice)
+		delete(resultStorage, res.ID)
+	}
 }
 
 // Stop the Katago Analyzer and the goroutines
 func (an *Analyzer) Stop() error {
-	glog.Infof("Shutting down Katago!")
+	glog.Infof("Shutting down Katago analyzer")
 	an.stdinQuit <- 1
 	an.collectorQuit <- 1
 	return nil
 }
 
 // Cmd creates the Katago analysis command.
-func (an *Analyzer) createCmd() {
-	threads := an.AnalysisThreads
-	an.cmd = exec.Command("katago", "analysis", "-model", an.Model, "-config", an.Config, "-analysis-threads", strconv.Itoa(threads))
+func (an *Analyzer) createCmd() *exec.Cmd {
+	return exec.Command("katago", "analysis", "-model", an.Model, "-config", an.Config, "-analysis-threads", strconv.Itoa(an.AnalysisThreads))
 }
 
-func (an *Analyzer) startAnalyzerIO() error {
+func (an *Analyzer) startAnalyzerIO(c *exec.Cmd) error {
```

Comment:
Why pass in the command vs referencing the command in the Analyzer? 
It _seems_ like this is a good practice because then you don't have to reason about the state of the Analyzer's command - there just is a command. Is that the right way to think about this?

Reply:
I was trying to reduce the amount of state that was being stored in the Analyzer --> Anything that's in the analyzer struct seems like it should live for the life of the object. The seems like it isn't like that -- it is used to things off and then we don't use the cmd again, since we quit by closing the STDIN Pipe.

I'm not tied to this change -- really just a matter of taste.